hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -19,86 +19,152 @@ object TableStage {
     globals: IR,
     partitioner: RVDPartitioner,
     contexts: IR,
-    body: (IR) => IR): TableStage = {
-    val globalsID = genUID()
-    new TableStage(
-      FastIndexedSeq(globalsID -> globals),
-      Set(globalsID),
-      Ref(globalsID, globals.typ),
+    body: (Ref) => IR
+  ): TableStage = {
+    val globalsRef = Ref(genUID(), globals.typ)
+    TableStage(
+      FastIndexedSeq(globalsRef.name -> globals),
+      FastIndexedSeq(globalsRef.name -> globalsRef),
+      globalsRef,
       partitioner,
-      contexts) {
-      def partition(ctxRef: Ref): IR = body(ctxRef)
-    }
+      contexts,
+      body)
+  }
+
+  def apply(
+    letBindings: IndexedSeq[(String, IR)],
+    broadcastVals: IndexedSeq[(String, IR)],
+    globals: Ref,
+    partitioner: RVDPartitioner,
+    contexts: IR,
+    partition: Ref => IR
+  ): TableStage = {
+    val ctxType = contexts.typ.asInstanceOf[TStream].elementType
+    val ctxRef = Ref(genUID(), ctxType)
+
+    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts, ctxRef.name, partition(ctxRef))
   }
 }
 
-abstract class TableStage(
+// Scope structure:
+// * 'letBindings' are evaluated in scope of previous 'letBindings', and are
+//   visible in 'broadcastVals' and 'contexts'.
+// * 'broadcastVals' are evaluated in scope of 'letBindings', and are visible
+//   in 'partitionIR'.
+// * 'globals' must be bound in 'letBindings', and rebound in 'broadcastVals',
+//   so 'globals' is visible both in later 'letBindings' and in 'partitionIR'.
+class TableStage(
   val letBindings: IndexedSeq[(String, IR)],
-  val broadcastVals: Set[String],
-  val globals: IR,
+  val broadcastVals: IndexedSeq[(String, IR)],
+  val globals: Ref,
   val partitioner: RVDPartitioner,
-  val contexts: IR) { self =>
+  val contexts: IR,
+  val ctxRefName: String,
+  val partitionIR: IR) { self =>
+
+  def ctxType: Type = contexts.typ.asInstanceOf[TStream].elementType
+  def rowType: TStruct = partitionIR.typ.asInstanceOf[TStream].elementType.asInstanceOf[TStruct]
+  def kType: TStruct = partitioner.kType
+  def key: IndexedSeq[String] = kType.fieldNames
+  def globalType: TStruct = globals.typ.asInstanceOf[TStruct]
+
+  assert(kType.fields.forall(f => rowType.field(f.name).typ == f.typ))
+  assert(broadcastVals.exists { case (name, value) => name == globals.name && value == globals})
+
+  def partition(ctx: IR): IR = {
+    require(ctx.typ == ctxType)
+    Let(ctxRefName, ctx, partitionIR)
+  }
 
-  def partition(ctxRef: Ref): IR
+  require(kType.fields.forall(f => rowType.field(f.name).typ == f.typ))
```

Comment:
This is redundant given the `assert` on line 71, no? Unless `require` and `assert` are different in a way I don't understand.

Reply:
require and assert just throw different exception types. Pretty similar for our purposes.

Reply:
Oops, yeah, completely redundant.

I think of require as stating preconditions for a function or class constructor, so a failure is the fault of the caller, while an assertion states an assumption about an internal state, so a failure is the fault of the immediate code. I like require for documentation purposes, explaining what a function assumes about its inputs in a self-checking way. But we've never discussed that as a team, and I think we tend to use assert for everything.